// firebase-tools/lib/database/removeRemote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request = require("request");
const responseToError = require("../responseToError");
const utils = require("../utils");
const FirebaseError = require("../error");
const logger = require("../logger");
const api = require("../api");
class RTDBRemoveRemote {
    constructor(instance) {
        this.instance = instance;
    }
    deletePath(path) {
        return this.patch(path, null, "all data");
    }
    deleteSubPath(path, subPaths) {
        const body = {};
        for (const c of subPaths) {
            body[c] = null;
        }
        return this.patch(path, body, `${subPaths.length} subpaths`);
    }
    listPath(path, numSubPath) {
        const url = utils.addSubdomain(api.realtimeOrigin, this.instance) +
            path +
            `.json?shallow=true&limitToFirst=${numSubPath}`;
        const t0 = Date.now();
        return api
            .addRequestHeaders({
            url,
        })
            .then((reqOptionsWithToken) => {
            return new Promise((resolve, reject) => {
                request.get(reqOptionsWithToken, (err, res, body) => {
                    if (err) {
                        return reject(new FirebaseError("Unexpected error while listing subtrees", {
                            exit: 2,
                            original: err,
                        }));
                    }
                    else if (res.statusCode >= 400) {
                        return reject(responseToError(res, body));
                    }
                    let data = {};
                    try {
                        data = JSON.parse(body);
                    }
                    catch (e) {
                        return reject(new FirebaseError("Malformed JSON response in shallow get ", {
                            exit: 2,
                            original: e,
                        }));
                    }
                    if (data) {
                        const keyList = Object.keys(data);
                        return resolve(keyList);
                    }
                    resolve([]);
                });
            });
        })
            .then((paths) => {
            const dt = Date.now() - t0;
            logger.debug(`[database] Sucessfully fetched ${paths.length} path at ${path} ${dt}`);
            return paths;
        });
    }
    patch(path, body, note) {
        const t0 = Date.now();
        return new Promise((resolve, reject) => {
            const url = utils.addSubdomain(api.realtimeOrigin, this.instance) +
                path +
                ".json?print=silent&writeSizeLimit=tiny";
            return api
                .addRequestHeaders({
                url,
                body,
                json: true,
            })
                .then((reqOptionsWithToken) => {
                request.patch(reqOptionsWithToken, (err, res, resBody) => {
                    if (err) {
                        return reject(new FirebaseError(`Unexpected error while removing data at ${path}`, {
                            exit: 2,
                            original: err,
                        }));
                    }
                    const dt = Date.now() - t0;
                    if (res.statusCode >= 400) {
                        logger.debug(`[database] Failed to remove ${note} at ${path} time: ${dt}`);
                        return resolve(false);
                    }
                    logger.debug(`[database] Sucessfully removed ${note} at ${path} time: ${dt}`);
                    return resolve(true);
                });
            });
        });
    }
}
exports.RTDBRemoveRemote = RTDBRemoveRemote;
