// firebase-tools/lib/test/database/fakeRemoveRemote.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pathLib = require("path");
const chai_1 = require("chai");
class FakeRemoveRemote {
    constructor(data, largeThreshold = 10) {
        this.data = data;
        this.largeThreshold = largeThreshold;
    }
    listPath(path, numChildren) {
        const d = this._dataAtpath(path);
        if (d) {
            const keys = Object.keys(d);
            return Promise.resolve(keys.slice(0, numChildren));
        }
        return Promise.resolve([]);
    }
    deletePath(path) {
        const d = this._dataAtpath(path);
        const size = this._size(d);
        if (size > this.largeThreshold) {
            return Promise.resolve(false);
        }
        this._deletePath(path);
        return Promise.resolve(true);
    }
    deleteSubPath(path, subPaths) {
        const d = this._dataAtpath(path);
        let size = 0;
        for (const p of subPaths) {
            size += this._size(d[p]);
        }
        if (size > this.largeThreshold) {
            return Promise.resolve(false);
        }
        for (const p of subPaths) {
            this.deletePath(`${path}/${p}`);
        }
        return Promise.resolve(true);
    }
    _deletePath(path) {
        if (path === "/") {
            this.data = null;
            return;
        }
        const parentDir = pathLib.dirname(path);
        const basename = pathLib.basename(path);
        delete this._dataAtpath(parentDir)[basename];
        if (Object.keys(this._dataAtpath(parentDir)).length === 0) {
            return this._deletePath(parentDir);
        }
    }
    _size(data) {
        if (typeof data === "number") {
            return data;
        }
        let size = 0;
        for (const key of Object.keys(data)) {
            size += this._size(data[key]);
        }
        return size;
    }
    _dataAtpath(path) {
        const splitedPath = path.slice(1).split("/");
        let d = this.data;
        for (const p of splitedPath) {
            if (d && p !== "") {
                if (typeof d === "number") {
                    d = null;
                }
                else {
                    d = d[p];
                }
            }
        }
        return d;
    }
}
exports.FakeRemoveRemote = FakeRemoveRemote;
describe("FakeRemoveRemote", () => {
    it("should return limit the number of subpaths returned", () => __awaiter(this, void 0, void 0, function* () {
        const fakeDb = new FakeRemoveRemote({ 1: 1, 2: 2, 3: 3, 4: 4 });
        yield chai_1.expect(fakeDb.listPath("/", 4)).to.eventually.eql(["1", "2", "3", "4"]);
        yield chai_1.expect(fakeDb.listPath("/", 3)).to.eventually.eql(["1", "2", "3"]);
        yield chai_1.expect(fakeDb.listPath("/", 2)).to.eventually.eql(["1", "2"]);
        yield chai_1.expect(fakeDb.listPath("/", 1)).to.eventually.eql(["1"]);
    }));
    it("should failed to delete large path /", () => __awaiter(this, void 0, void 0, function* () {
        const data = { 1: 11 };
        const fakeDb = new FakeRemoveRemote(data);
        yield chai_1.expect(fakeDb.deletePath("/")).to.eventually.eql(false);
        chai_1.expect(fakeDb.data).eql(data);
    }));
    it("should sucessfully delete large path /", () => __awaiter(this, void 0, void 0, function* () {
        const fakeDb = new FakeRemoveRemote({ 1: 9 });
        yield chai_1.expect(fakeDb.deletePath("/")).to.eventually.eql(true);
        chai_1.expect(fakeDb.data).eql(null);
    }));
    it("should failed to delete large path /1", () => __awaiter(this, void 0, void 0, function* () {
        const data = { 1: { a: 3, b: 9, c: 2, d: 3 } };
        const fakeDb = new FakeRemoveRemote(data);
        yield chai_1.expect(fakeDb.deletePath("/1")).to.eventually.eql(false);
        chai_1.expect(fakeDb.data).eql(data);
    }));
    it("should successfully delete path /1/a", () => __awaiter(this, void 0, void 0, function* () {
        const fakeDb = new FakeRemoveRemote({ 1: { a: 3, b: 9, c: 2, d: 3 } });
        yield chai_1.expect(fakeDb.deletePath("/1/a")).to.eventually.eql(true);
        chai_1.expect(fakeDb.data).eql({ 1: { b: 9, c: 2, d: 3 } });
    }));
    it("should failed to delete large paths /1/a /1/b", () => __awaiter(this, void 0, void 0, function* () {
        const data = { 1: { a: 3, b: 9, c: 2, d: 3 } };
        const fakeDb = new FakeRemoveRemote(data);
        yield chai_1.expect(fakeDb.deleteSubPath("/1", ["a", "b"])).to.eventually.eql(false);
        chai_1.expect(fakeDb.data).eql(data);
    }));
    it("should successfully delete multi paths /1/c /1/d", () => __awaiter(this, void 0, void 0, function* () {
        const fakeDb = new FakeRemoveRemote({ 1: { a: 3, b: 9, c: 2, d: 3 } });
        yield chai_1.expect(fakeDb.deleteSubPath("/1", ["c", "d"])).to.eventually.eql(true);
        chai_1.expect(fakeDb.data).eql({ 1: { a: 3, b: 9 } });
    }));
});
