// firebase-tools/lib/test/database/remove.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const remove_1 = require("../../database/remove");
const fakeRemoveRemote_spec_1 = require("./fakeRemoveRemote.spec");
describe("DatabaseRemove", () => {
    it("should remove tiny tree", () => __awaiter(this, void 0, void 0, function* () {
        const fakeDb = new fakeRemoveRemote_spec_1.FakeRemoveRemote({ c: 1 });
        const removeOps = new remove_1.default("test-tiny-tree", "/");
        removeOps.remote = fakeDb;
        yield removeOps.execute();
        chai_1.expect(fakeDb.data).to.eql(null);
    }));
    it("should remove subtree at /a/b/c", () => __awaiter(this, void 0, void 0, function* () {
        const fakeDb = new fakeRemoveRemote_spec_1.FakeRemoveRemote({
            a: {
                b: { x: { y: 1 } },
                c: { x: 4, y: 8 },
                d: 10,
            },
            d: {
                e: 3,
            },
        });
        const removeOps = new remove_1.default("test-sub-path", "/a");
        removeOps.remote = fakeDb;
        yield removeOps.execute();
        chai_1.expect(fakeDb.data).to.eql({
            d: {
                e: 3,
            },
        });
    }));
    function buildData(branchFactor, depth) {
        if (depth === 0) {
            return 1;
        }
        const d = {};
        for (let i = 0; i < branchFactor; i++) {
            d[`${i}`] = buildData(branchFactor, depth - 1);
        }
        return d;
    }
    function databaseRemoveTestSuit(threshold) {
        describe(`DatabaseRemove when largeThreshold=${threshold}`, () => {
            it("should remove nested tree", () => __awaiter(this, void 0, void 0, function* () {
                const fakeDb = new fakeRemoveRemote_spec_1.FakeRemoveRemote(buildData(3, 5), threshold);
                const removeOps = new remove_1.default("test-nested-tree", "/");
                removeOps.remote = fakeDb;
                yield removeOps.execute();
                chai_1.expect(fakeDb.data).to.eql(null);
            }));
            it("should remove flat tree when threshold=${threshold}", () => __awaiter(this, void 0, void 0, function* () {
                const fakeDb = new fakeRemoveRemote_spec_1.FakeRemoveRemote(buildData(1232, 1), threshold);
                const removeOps = new remove_1.default("test-remover", "/");
                removeOps.remote = fakeDb;
                yield removeOps.execute();
                chai_1.expect(fakeDb.data).to.eql(null);
            }));
        });
    }
    databaseRemoveTestSuit(100);
    databaseRemoveTestSuit(10);
    databaseRemoveTestSuit(1);
});
