// googleapis/build/src/apis/index.js
"use strict";
// Copyright 2012-2016, Google, Inc.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
/*! THIS FILE IS AUTO-GENERATED */
var path = require("path");
var util = require("util");
var APIS = {
    acceleratedmobilepageurl: {
        'v1': require('./acceleratedmobilepageurl/v1'),
    },
    adexchangebuyer: {
        'v1.2': require('./adexchangebuyer/v1.2'),
        'v1.3': require('./adexchangebuyer/v1.3'),
        'v1.4': require('./adexchangebuyer/v1.4'),
    },
    adexchangebuyer2: {
        'v2beta1': require('./adexchangebuyer2/v2beta1'),
    },
    adexchangeseller: {
        'v1.1': require('./adexchangeseller/v1.1'),
        'v1': require('./adexchangeseller/v1'),
        'v2.0': require('./adexchangeseller/v2.0'),
    },
    adexperiencereport: {
        'v1': require('./adexperiencereport/v1'),
    },
    admin: {
        'datatransfer_v1': require('./admin/datatransfer_v1'),
        'directory_v1': require('./admin/directory_v1'),
        'reports_v1': require('./admin/reports_v1'),
    },
    adsense: {
        'v1.4': require('./adsense/v1.4'),
    },
    adsensehost: {
        'v4.1': require('./adsensehost/v4.1'),
    },
    analytics: {
        'v2.4': require('./analytics/v2.4'),
        'v3': require('./analytics/v3'),
    },
    analyticsreporting: {
        'v4': require('./analyticsreporting/v4'),
    },
    androiddeviceprovisioning: {
        'v1': require('./androiddeviceprovisioning/v1'),
    },
    androidenterprise: {
        'v1': require('./androidenterprise/v1'),
    },
    androidmanagement: {
        'v1': require('./androidmanagement/v1'),
    },
    androidpublisher: {
        'v1.1': require('./androidpublisher/v1.1'),
        'v1': require('./androidpublisher/v1'),
        'v2': require('./androidpublisher/v2'),
    },
    appengine: {
        'v1': require('./appengine/v1'),
        'v1alpha': require('./appengine/v1alpha'),
        'v1beta': require('./appengine/v1beta'),
        'v1beta4': require('./appengine/v1beta4'),
        'v1beta5': require('./appengine/v1beta5'),
    },
    appsactivity: {
        'v1': require('./appsactivity/v1'),
    },
    appstate: {
        'v1': require('./appstate/v1'),
    },
    bigquery: {
        'v2': require('./bigquery/v2'),
    },
    bigquerydatatransfer: {
        'v1': require('./bigquerydatatransfer/v1'),
    },
    blogger: {
        'v2': require('./blogger/v2'),
        'v3': require('./blogger/v3'),
    },
    books: {
        'v1': require('./books/v1'),
    },
    calendar: {
        'v3': require('./calendar/v3'),
    },
    civicinfo: {
        'v2': require('./civicinfo/v2'),
    },
    classroom: {
        'v1': require('./classroom/v1'),
    },
    cloudbilling: {
        'v1': require('./cloudbilling/v1'),
    },
    cloudbuild: {
        'v1': require('./cloudbuild/v1'),
    },
    clouddebugger: {
        'v2': require('./clouddebugger/v2'),
    },
    clouderrorreporting: {
        'v1beta1': require('./clouderrorreporting/v1beta1'),
    },
    cloudfunctions: {
        'v1': require('./cloudfunctions/v1'),
        'v1beta2': require('./cloudfunctions/v1beta2'),
    },
    cloudiot: {
        'v1': require('./cloudiot/v1'),
        'v1beta1': require('./cloudiot/v1beta1'),
    },
    cloudkms: {
        'v1': require('./cloudkms/v1'),
    },
    cloudresourcemanager: {
        'v1': require('./cloudresourcemanager/v1'),
        'v1beta1': require('./cloudresourcemanager/v1beta1'),
        'v2beta1': require('./cloudresourcemanager/v2beta1'),
    },
    cloudtrace: {
        'v1': require('./cloudtrace/v1'),
        'v2': require('./cloudtrace/v2'),
    },
    clouduseraccounts: {
        'alpha': require('./clouduseraccounts/alpha'),
        'beta': require('./clouduseraccounts/beta'),
        'vm_alpha': require('./clouduseraccounts/vm_alpha'),
        'vm_beta': require('./clouduseraccounts/vm_beta'),
    },
    compute: {
        'alpha': require('./compute/alpha'),
        'beta': require('./compute/beta'),
        'v1': require('./compute/v1'),
    },
    consumersurveys: {
        'v2': require('./consumersurveys/v2'),
    },
    container: {
        'v1': require('./container/v1'),
        'v1beta1': require('./container/v1beta1'),
    },
    content: {
        'v2': require('./content/v2'),
        'v2sandbox': require('./content/v2sandbox'),
    },
    customsearch: {
        'v1': require('./customsearch/v1'),
    },
    dataflow: {
        'v1b3': require('./dataflow/v1b3'),
    },
    dataproc: {
        'v1': require('./dataproc/v1'),
        'v1beta2': require('./dataproc/v1beta2'),
    },
    datastore: {
        'v1': require('./datastore/v1'),
        'v1beta1': require('./datastore/v1beta1'),
        'v1beta3': require('./datastore/v1beta3'),
    },
    deploymentmanager: {
        'alpha': require('./deploymentmanager/alpha'),
        'v2': require('./deploymentmanager/v2'),
        'v2beta': require('./deploymentmanager/v2beta'),
    },
    dfareporting: {
        'v2.7': require('./dfareporting/v2.7'),
        'v2.8': require('./dfareporting/v2.8'),
        'v3.0': require('./dfareporting/v3.0'),
    },
    dialogflow: {
        'v2beta1': require('./dialogflow/v2beta1'),
    },
    discovery: {
        'v1': require('./discovery/v1'),
    },
    dlp: {
        'v2beta1': require('./dlp/v2beta1'),
    },
    dns: {
        'v1': require('./dns/v1'),
        'v1beta2': require('./dns/v1beta2'),
        'v2beta1': require('./dns/v2beta1'),
    },
    doubleclickbidmanager: {
        'v1': require('./doubleclickbidmanager/v1'),
    },
    doubleclicksearch: {
        'v2': require('./doubleclicksearch/v2'),
    },
    drive: {
        'v2': require('./drive/v2'),
        'v3': require('./drive/v3'),
    },
    firebasedynamiclinks: {
        'v1': require('./firebasedynamiclinks/v1'),
    },
    firebaseremoteconfig: {
        'v1': require('./firebaseremoteconfig/v1'),
    },
    firebaserules: {
        'v1': require('./firebaserules/v1'),
    },
    firestore: {
        'v1beta1': require('./firestore/v1beta1'),
    },
    fitness: {
        'v1': require('./fitness/v1'),
    },
    fusiontables: {
        'v1': require('./fusiontables/v1'),
        'v2': require('./fusiontables/v2'),
    },
    games: {
        'v1': require('./games/v1'),
    },
    gamesConfiguration: {
        'v1configuration': require('./gamesConfiguration/v1configuration'),
    },
    gamesManagement: {
        'v1management': require('./gamesManagement/v1management'),
    },
    genomics: {
        'v1': require('./genomics/v1'),
        'v1alpha2': require('./genomics/v1alpha2'),
    },
    gmail: {
        'v1': require('./gmail/v1'),
    },
    groupsmigration: {
        'v1': require('./groupsmigration/v1'),
    },
    groupssettings: {
        'v1': require('./groupssettings/v1'),
    },
    iam: {
        'v1': require('./iam/v1'),
    },
    identitytoolkit: {
        'v3': require('./identitytoolkit/v3'),
    },
    kgsearch: {
        'v1': require('./kgsearch/v1'),
    },
    language: {
        'v1': require('./language/v1'),
        'v1beta1': require('./language/v1beta1'),
        'v1beta2': require('./language/v1beta2'),
    },
    licensing: {
        'v1': require('./licensing/v1'),
    },
    logging: {
        'v2': require('./logging/v2'),
        'v2beta1': require('./logging/v2beta1'),
    },
    manufacturers: {
        'v1': require('./manufacturers/v1'),
    },
    mirror: {
        'v1': require('./mirror/v1'),
    },
    ml: {
        'v1': require('./ml/v1'),
    },
    monitoring: {
        'v3': require('./monitoring/v3'),
    },
    oauth2: {
        'v1': require('./oauth2/v1'),
        'v2': require('./oauth2/v2'),
    },
    oslogin: {
        'v1': require('./oslogin/v1'),
        'v1alpha': require('./oslogin/v1alpha'),
        'v1beta': require('./oslogin/v1beta'),
    },
    pagespeedonline: {
        'v1': require('./pagespeedonline/v1'),
        'v2': require('./pagespeedonline/v2'),
    },
    partners: {
        'v2': require('./partners/v2'),
    },
    people: {
        'v1': require('./people/v1'),
    },
    playcustomapp: {
        'v1': require('./playcustomapp/v1'),
    },
    plus: {
        'v1': require('./plus/v1'),
    },
    plusDomains: {
        'v1': require('./plusDomains/v1'),
    },
    prediction: {
        'v1.2': require('./prediction/v1.2'),
        'v1.3': require('./prediction/v1.3'),
        'v1.4': require('./prediction/v1.4'),
        'v1.5': require('./prediction/v1.5'),
        'v1.6': require('./prediction/v1.6'),
    },
    proximitybeacon: {
        'v1beta1': require('./proximitybeacon/v1beta1'),
    },
    pubsub: {
        'v1': require('./pubsub/v1'),
        'v1beta1a': require('./pubsub/v1beta1a'),
        'v1beta2': require('./pubsub/v1beta2'),
    },
    replicapool: {
        'v1beta1': require('./replicapool/v1beta1'),
        'v1beta2': require('./replicapool/v1beta2'),
    },
    replicapoolupdater: {
        'v1beta1': require('./replicapoolupdater/v1beta1'),
    },
    reseller: {
        'v1': require('./reseller/v1'),
    },
    resourceviews: {
        'v1beta1': require('./resourceviews/v1beta1'),
        'v1beta2': require('./resourceviews/v1beta2'),
    },
    runtimeconfig: {
        'v1': require('./runtimeconfig/v1'),
        'v1beta1': require('./runtimeconfig/v1beta1'),
    },
    safebrowsing: {
        'v4': require('./safebrowsing/v4'),
    },
    script: {
        'v1': require('./script/v1'),
    },
    searchconsole: {
        'v1': require('./searchconsole/v1'),
    },
    serviceconsumermanagement: {
        'v1': require('./serviceconsumermanagement/v1'),
    },
    servicecontrol: {
        'v1': require('./servicecontrol/v1'),
    },
    servicemanagement: {
        'v1': require('./servicemanagement/v1'),
    },
    sheets: {
        'v4': require('./sheets/v4'),
    },
    siteVerification: {
        'v1': require('./siteVerification/v1'),
    },
    slides: {
        'v1': require('./slides/v1'),
    },
    sourcerepo: {
        'v1': require('./sourcerepo/v1'),
    },
    spanner: {
        'v1': require('./spanner/v1'),
    },
    spectrum: {
        'v1explorer': require('./spectrum/v1explorer'),
    },
    speech: {
        'v1': require('./speech/v1'),
        'v1beta1': require('./speech/v1beta1'),
    },
    sqladmin: {
        'v1beta3': require('./sqladmin/v1beta3'),
        'v1beta4': require('./sqladmin/v1beta4'),
    },
    storage: {
        'v1': require('./storage/v1'),
        'v1beta1': require('./storage/v1beta1'),
        'v1beta2': require('./storage/v1beta2'),
    },
    storagetransfer: {
        'v1': require('./storagetransfer/v1'),
    },
    streetviewpublish: {
        'v1': require('./streetviewpublish/v1'),
    },
    surveys: {
        'v2': require('./surveys/v2'),
    },
    tagmanager: {
        'v1': require('./tagmanager/v1'),
        'v2': require('./tagmanager/v2'),
    },
    taskqueue: {
        'v1beta1': require('./taskqueue/v1beta1'),
        'v1beta2': require('./taskqueue/v1beta2'),
    },
    tasks: {
        'v1': require('./tasks/v1'),
    },
    testing: {
        'v1': require('./testing/v1'),
    },
    toolresults: {
        'v1beta3': require('./toolresults/v1beta3'),
    },
    tpu: {
        'v1alpha1': require('./tpu/v1alpha1'),
    },
    translate: {
        'v2': require('./translate/v2'),
    },
    urlshortener: {
        'v1': require('./urlshortener/v1'),
    },
    vault: {
        'v1': require('./vault/v1'),
    },
    videointelligence: {
        'v1beta1': require('./videointelligence/v1beta1'),
    },
    vision: {
        'v1': require('./vision/v1'),
        'v1p1beta1': require('./vision/v1p1beta1'),
    },
    webfonts: {
        'v1': require('./webfonts/v1'),
    },
    webmasters: {
        'v3': require('./webmasters/v3'),
    },
    youtube: {
        'v3': require('./youtube/v3'),
    },
    youtubeAnalytics: {
        'v1': require('./youtubeAnalytics/v1'),
        'v1beta1': require('./youtubeAnalytics/v1beta1'),
    },
    youtubereporting: {
        'v1': require('./youtubereporting/v1'),
    },
};
function getAPI(api, options) {
    var type = typeof options;
    var version;
    if (type === 'string') {
        version = options;
        options = {};
    }
    else if (type === 'object') {
        version = options.version;
        delete options.version;
    }
    else {
        throw new Error('Argument error: Accepts only string or object');
    }
    try {
        var endpoint = APIS[api][path.basename(version)];
        var ep = new endpoint(options);
        ep.google = this; // for drive.google.transporter
        return Object.freeze(ep); // create new & freeze
    }
    catch (e) {
        throw new Error(util.format('Unable to load endpoint %s("%s"): %s', api, version, e.message));
    }
}
function acceleratedmobilepageurl(options) {
    return getAPI.call(this, 'acceleratedmobilepageurl', options);
}
exports.acceleratedmobilepageurl = acceleratedmobilepageurl;
function adexchangebuyer(options) {
    return getAPI.call(this, 'adexchangebuyer', options);
}
exports.adexchangebuyer = adexchangebuyer;
function adexchangebuyer2(options) {
    return getAPI.call(this, 'adexchangebuyer2', options);
}
exports.adexchangebuyer2 = adexchangebuyer2;
function adexchangeseller(options) {
    return getAPI.call(this, 'adexchangeseller', options);
}
exports.adexchangeseller = adexchangeseller;
function adexperiencereport(options) {
    return getAPI.call(this, 'adexperiencereport', options);
}
exports.adexperiencereport = adexperiencereport;
function admin(options) {
    return getAPI.call(this, 'admin', options);
}
exports.admin = admin;
function adsense(options) {
    return getAPI.call(this, 'adsense', options);
}
exports.adsense = adsense;
function adsensehost(options) {
    return getAPI.call(this, 'adsensehost', options);
}
exports.adsensehost = adsensehost;
function analytics(options) {
    return getAPI.call(this, 'analytics', options);
}
exports.analytics = analytics;
function analyticsreporting(options) {
    return getAPI.call(this, 'analyticsreporting', options);
}
exports.analyticsreporting = analyticsreporting;
function androiddeviceprovisioning(options) {
    return getAPI.call(this, 'androiddeviceprovisioning', options);
}
exports.androiddeviceprovisioning = androiddeviceprovisioning;
function androidenterprise(options) {
    return getAPI.call(this, 'androidenterprise', options);
}
exports.androidenterprise = androidenterprise;
function androidmanagement(options) {
    return getAPI.call(this, 'androidmanagement', options);
}
exports.androidmanagement = androidmanagement;
function androidpublisher(options) {
    return getAPI.call(this, 'androidpublisher', options);
}
exports.androidpublisher = androidpublisher;
function appengine(options) {
    return getAPI.call(this, 'appengine', options);
}
exports.appengine = appengine;
function appsactivity(options) {
    return getAPI.call(this, 'appsactivity', options);
}
exports.appsactivity = appsactivity;
function appstate(options) {
    return getAPI.call(this, 'appstate', options);
}
exports.appstate = appstate;
function bigquery(options) {
    return getAPI.call(this, 'bigquery', options);
}
exports.bigquery = bigquery;
function bigquerydatatransfer(options) {
    return getAPI.call(this, 'bigquerydatatransfer', options);
}
exports.bigquerydatatransfer = bigquerydatatransfer;
function blogger(options) {
    return getAPI.call(this, 'blogger', options);
}
exports.blogger = blogger;
function books(options) {
    return getAPI.call(this, 'books', options);
}
exports.books = books;
function calendar(options) {
    return getAPI.call(this, 'calendar', options);
}
exports.calendar = calendar;
function civicinfo(options) {
    return getAPI.call(this, 'civicinfo', options);
}
exports.civicinfo = civicinfo;
function classroom(options) {
    return getAPI.call(this, 'classroom', options);
}
exports.classroom = classroom;
function cloudbilling(options) {
    return getAPI.call(this, 'cloudbilling', options);
}
exports.cloudbilling = cloudbilling;
function cloudbuild(options) {
    return getAPI.call(this, 'cloudbuild', options);
}
exports.cloudbuild = cloudbuild;
function clouddebugger(options) {
    return getAPI.call(this, 'clouddebugger', options);
}
exports.clouddebugger = clouddebugger;
function clouderrorreporting(options) {
    return getAPI.call(this, 'clouderrorreporting', options);
}
exports.clouderrorreporting = clouderrorreporting;
function cloudfunctions(options) {
    return getAPI.call(this, 'cloudfunctions', options);
}
exports.cloudfunctions = cloudfunctions;
function cloudiot(options) {
    return getAPI.call(this, 'cloudiot', options);
}
exports.cloudiot = cloudiot;
function cloudkms(options) {
    return getAPI.call(this, 'cloudkms', options);
}
exports.cloudkms = cloudkms;
function cloudresourcemanager(options) {
    return getAPI.call(this, 'cloudresourcemanager', options);
}
exports.cloudresourcemanager = cloudresourcemanager;
function cloudtrace(options) {
    return getAPI.call(this, 'cloudtrace', options);
}
exports.cloudtrace = cloudtrace;
function clouduseraccounts(options) {
    return getAPI.call(this, 'clouduseraccounts', options);
}
exports.clouduseraccounts = clouduseraccounts;
function compute(options) {
    return getAPI.call(this, 'compute', options);
}
exports.compute = compute;
function consumersurveys(options) {
    return getAPI.call(this, 'consumersurveys', options);
}
exports.consumersurveys = consumersurveys;
function container(options) {
    return getAPI.call(this, 'container', options);
}
exports.container = container;
function content(options) {
    return getAPI.call(this, 'content', options);
}
exports.content = content;
function customsearch(options) {
    return getAPI.call(this, 'customsearch', options);
}
exports.customsearch = customsearch;
function dataflow(options) {
    return getAPI.call(this, 'dataflow', options);
}
exports.dataflow = dataflow;
function dataproc(options) {
    return getAPI.call(this, 'dataproc', options);
}
exports.dataproc = dataproc;
function datastore(options) {
    return getAPI.call(this, 'datastore', options);
}
exports.datastore = datastore;
function deploymentmanager(options) {
    return getAPI.call(this, 'deploymentmanager', options);
}
exports.deploymentmanager = deploymentmanager;
function dfareporting(options) {
    return getAPI.call(this, 'dfareporting', options);
}
exports.dfareporting = dfareporting;
function dialogflow(options) {
    return getAPI.call(this, 'dialogflow', options);
}
exports.dialogflow = dialogflow;
function discovery(options) {
    return getAPI.call(this, 'discovery', options);
}
exports.discovery = discovery;
function dlp(options) {
    return getAPI.call(this, 'dlp', options);
}
exports.dlp = dlp;
function dns(options) {
    return getAPI.call(this, 'dns', options);
}
exports.dns = dns;
function doubleclickbidmanager(options) {
    return getAPI.call(this, 'doubleclickbidmanager', options);
}
exports.doubleclickbidmanager = doubleclickbidmanager;
function doubleclicksearch(options) {
    return getAPI.call(this, 'doubleclicksearch', options);
}
exports.doubleclicksearch = doubleclicksearch;
function drive(options) {
    return getAPI.call(this, 'drive', options);
}
exports.drive = drive;
function firebasedynamiclinks(options) {
    return getAPI.call(this, 'firebasedynamiclinks', options);
}
exports.firebasedynamiclinks = firebasedynamiclinks;
function firebaseremoteconfig(options) {
    return getAPI.call(this, 'firebaseremoteconfig', options);
}
exports.firebaseremoteconfig = firebaseremoteconfig;
function firebaserules(options) {
    return getAPI.call(this, 'firebaserules', options);
}
exports.firebaserules = firebaserules;
function firestore(options) {
    return getAPI.call(this, 'firestore', options);
}
exports.firestore = firestore;
function fitness(options) {
    return getAPI.call(this, 'fitness', options);
}
exports.fitness = fitness;
function fusiontables(options) {
    return getAPI.call(this, 'fusiontables', options);
}
exports.fusiontables = fusiontables;
function games(options) {
    return getAPI.call(this, 'games', options);
}
exports.games = games;
function gamesConfiguration(options) {
    return getAPI.call(this, 'gamesConfiguration', options);
}
exports.gamesConfiguration = gamesConfiguration;
function gamesManagement(options) {
    return getAPI.call(this, 'gamesManagement', options);
}
exports.gamesManagement = gamesManagement;
function genomics(options) {
    return getAPI.call(this, 'genomics', options);
}
exports.genomics = genomics;
function gmail(options) {
    return getAPI.call(this, 'gmail', options);
}
exports.gmail = gmail;
function groupsmigration(options) {
    return getAPI.call(this, 'groupsmigration', options);
}
exports.groupsmigration = groupsmigration;
function groupssettings(options) {
    return getAPI.call(this, 'groupssettings', options);
}
exports.groupssettings = groupssettings;
function iam(options) {
    return getAPI.call(this, 'iam', options);
}
exports.iam = iam;
function identitytoolkit(options) {
    return getAPI.call(this, 'identitytoolkit', options);
}
exports.identitytoolkit = identitytoolkit;
function kgsearch(options) {
    return getAPI.call(this, 'kgsearch', options);
}
exports.kgsearch = kgsearch;
function language(options) {
    return getAPI.call(this, 'language', options);
}
exports.language = language;
function licensing(options) {
    return getAPI.call(this, 'licensing', options);
}
exports.licensing = licensing;
function logging(options) {
    return getAPI.call(this, 'logging', options);
}
exports.logging = logging;
function manufacturers(options) {
    return getAPI.call(this, 'manufacturers', options);
}
exports.manufacturers = manufacturers;
function mirror(options) {
    return getAPI.call(this, 'mirror', options);
}
exports.mirror = mirror;
function ml(options) {
    return getAPI.call(this, 'ml', options);
}
exports.ml = ml;
function monitoring(options) {
    return getAPI.call(this, 'monitoring', options);
}
exports.monitoring = monitoring;
function oauth2(options) {
    return getAPI.call(this, 'oauth2', options);
}
exports.oauth2 = oauth2;
function oslogin(options) {
    return getAPI.call(this, 'oslogin', options);
}
exports.oslogin = oslogin;
function pagespeedonline(options) {
    return getAPI.call(this, 'pagespeedonline', options);
}
exports.pagespeedonline = pagespeedonline;
function partners(options) {
    return getAPI.call(this, 'partners', options);
}
exports.partners = partners;
function people(options) {
    return getAPI.call(this, 'people', options);
}
exports.people = people;
function playcustomapp(options) {
    return getAPI.call(this, 'playcustomapp', options);
}
exports.playcustomapp = playcustomapp;
function plus(options) {
    return getAPI.call(this, 'plus', options);
}
exports.plus = plus;
function plusDomains(options) {
    return getAPI.call(this, 'plusDomains', options);
}
exports.plusDomains = plusDomains;
function prediction(options) {
    return getAPI.call(this, 'prediction', options);
}
exports.prediction = prediction;
function proximitybeacon(options) {
    return getAPI.call(this, 'proximitybeacon', options);
}
exports.proximitybeacon = proximitybeacon;
function pubsub(options) {
    return getAPI.call(this, 'pubsub', options);
}
exports.pubsub = pubsub;
function replicapool(options) {
    return getAPI.call(this, 'replicapool', options);
}
exports.replicapool = replicapool;
function replicapoolupdater(options) {
    return getAPI.call(this, 'replicapoolupdater', options);
}
exports.replicapoolupdater = replicapoolupdater;
function reseller(options) {
    return getAPI.call(this, 'reseller', options);
}
exports.reseller = reseller;
function resourceviews(options) {
    return getAPI.call(this, 'resourceviews', options);
}
exports.resourceviews = resourceviews;
function runtimeconfig(options) {
    return getAPI.call(this, 'runtimeconfig', options);
}
exports.runtimeconfig = runtimeconfig;
function safebrowsing(options) {
    return getAPI.call(this, 'safebrowsing', options);
}
exports.safebrowsing = safebrowsing;
function script(options) {
    return getAPI.call(this, 'script', options);
}
exports.script = script;
function searchconsole(options) {
    return getAPI.call(this, 'searchconsole', options);
}
exports.searchconsole = searchconsole;
function serviceconsumermanagement(options) {
    return getAPI.call(this, 'serviceconsumermanagement', options);
}
exports.serviceconsumermanagement = serviceconsumermanagement;
function servicecontrol(options) {
    return getAPI.call(this, 'servicecontrol', options);
}
exports.servicecontrol = servicecontrol;
function servicemanagement(options) {
    return getAPI.call(this, 'servicemanagement', options);
}
exports.servicemanagement = servicemanagement;
function sheets(options) {
    return getAPI.call(this, 'sheets', options);
}
exports.sheets = sheets;
function siteVerification(options) {
    return getAPI.call(this, 'siteVerification', options);
}
exports.siteVerification = siteVerification;
function slides(options) {
    return getAPI.call(this, 'slides', options);
}
exports.slides = slides;
function sourcerepo(options) {
    return getAPI.call(this, 'sourcerepo', options);
}
exports.sourcerepo = sourcerepo;
function spanner(options) {
    return getAPI.call(this, 'spanner', options);
}
exports.spanner = spanner;
function spectrum(options) {
    return getAPI.call(this, 'spectrum', options);
}
exports.spectrum = spectrum;
function speech(options) {
    return getAPI.call(this, 'speech', options);
}
exports.speech = speech;
function sqladmin(options) {
    return getAPI.call(this, 'sqladmin', options);
}
exports.sqladmin = sqladmin;
function storage(options) {
    return getAPI.call(this, 'storage', options);
}
exports.storage = storage;
function storagetransfer(options) {
    return getAPI.call(this, 'storagetransfer', options);
}
exports.storagetransfer = storagetransfer;
function streetviewpublish(options) {
    return getAPI.call(this, 'streetviewpublish', options);
}
exports.streetviewpublish = streetviewpublish;
function surveys(options) {
    return getAPI.call(this, 'surveys', options);
}
exports.surveys = surveys;
function tagmanager(options) {
    return getAPI.call(this, 'tagmanager', options);
}
exports.tagmanager = tagmanager;
function taskqueue(options) {
    return getAPI.call(this, 'taskqueue', options);
}
exports.taskqueue = taskqueue;
function tasks(options) {
    return getAPI.call(this, 'tasks', options);
}
exports.tasks = tasks;
function testing(options) {
    return getAPI.call(this, 'testing', options);
}
exports.testing = testing;
function toolresults(options) {
    return getAPI.call(this, 'toolresults', options);
}
exports.toolresults = toolresults;
function tpu(options) {
    return getAPI.call(this, 'tpu', options);
}
exports.tpu = tpu;
function translate(options) {
    return getAPI.call(this, 'translate', options);
}
exports.translate = translate;
function urlshortener(options) {
    return getAPI.call(this, 'urlshortener', options);
}
exports.urlshortener = urlshortener;
function vault(options) {
    return getAPI.call(this, 'vault', options);
}
exports.vault = vault;
function videointelligence(options) {
    return getAPI.call(this, 'videointelligence', options);
}
exports.videointelligence = videointelligence;
function vision(options) {
    return getAPI.call(this, 'vision', options);
}
exports.vision = vision;
function webfonts(options) {
    return getAPI.call(this, 'webfonts', options);
}
exports.webfonts = webfonts;
function webmasters(options) {
    return getAPI.call(this, 'webmasters', options);
}
exports.webmasters = webmasters;
function youtube(options) {
    return getAPI.call(this, 'youtube', options);
}
exports.youtube = youtube;
function youtubeAnalytics(options) {
    return getAPI.call(this, 'youtubeAnalytics', options);
}
exports.youtubeAnalytics = youtubeAnalytics;
function youtubereporting(options) {
    return getAPI.call(this, 'youtubereporting', options);
}
exports.youtubereporting = youtubereporting;
